### PR TITLE
Add Admin Route

### PR DESCRIPTION
**Pull request by Bino26**
Admin Route may help us to change the user role as admin .
We need this because only admin can add and delete quizzes .

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,4 @@
+PORT=8000
```
Comment (in43sh) [inline .env.example:1]: For security reasons, I would suggest to not use the real backend port.

File: .env.example [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+MONGO_URI=mongodb+srv://...cluster
```
Comment (in43sh) [inline .env.example:2]: I've seen it before that MONGO_URI in .env.example is written like `mongodb+srv://username:password@cluster.example.com/mydatabase`.

File: .env.example [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+SECRET=SECRET
```
Comment (in43sh) [inline .env.example:3]: Again for security reasons, I would not exposed a real secret key for JWT token. You can replace it with `YOUR_SECRET_HERE`.

File: .env.example [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+CLIENT_URL=http://localhost:3000
```
Comment (in43sh) [inline .env.example:4]: This may or may not be sensitive, but I would rather avoid using a real port from production in a publicly available file.

File: src/controllers/userController.js [hunk 5]
```diff
@@ -73,6 +73,62 @@ const logIn = async (req, res) => {
   }
 };
 
+/******************************************************
+ * @GETUSER
+ * @route /api/v1/login
+ * @method GET
+ * @description retrieve user data from mongoDb if user is valid(jwt auth)
```
Comment (in43sh) [inline src/controllers/userController.js:80]: This description is confusing. Could you rephrase it?

File: src/controllers/userController.js [hunk 6]
```diff
@@ -73,6 +73,62 @@ const logIn = async (req, res) => {
...
+ * @returns User Object
```
Comment (in43sh) [inline src/controllers/userController.js:81]: I checked and the response has a different format:
```
{
    "success": true,
    "message": "User data got  sucessfully",
    "data": {
        "userId": "",
        "email": "test@test.com",
        "firstname": "test",
        "lastname": "test",
        "role": "admin",
        "accessToken": ""
    }
}
```
I removed `userId` and `accessToken`.

File: src/controllers/userController.js [hunk 7]
```diff
@@ -73,6 +73,62 @@ const logIn = async (req, res) => {
...
+ ******************************************************/
+
+const getUser = async (req, res) => {
+  const { userId } = req.user;
+  try {
+    const user = await userModel.findById(userId);
+    return res.status(200).json({
+      success: true,
+      message: "User data got  sucessfully",
```
Comment (in43sh) [inline src/controllers/userController.js:90]: Please, remove an extra space. Also, I would replace it with `User data retrieved successfully`.

File: src/controllers/userController.js [hunk 8]
```diff
@@ -73,6 +73,62 @@ const logIn = async (req, res) => {
...
+    return res.status(400).json({
+      success: false,
+      message: error.message,
+    });
+  }
+};
+/******************************************************
+ * @GETADMIN
+ * @route /api/v1/admin
+ * @method GET
```
Comment (in43sh) [inline src/controllers/userController.js:103]: In the application, you are using PUT method, so I would change it here too.

File: src/controllers/userController.js [hunk 9]
```diff
@@ -73,6 +73,62 @@ const logIn = async (req, res) => {
...
+ * @description retrieve user data from mongoDb if user is valid(jwt auth)
```
Comment (in43sh) [inline src/controllers/userController.js:104]: The description here is the same as in the previous method. Could you update it?

File: src/controllers/userController.js [hunk 10]
```diff
@@ -73,6 +73,62 @@ const logIn = async (req, res) => {
...
+ * @returns User Object
```
Comment (in43sh) [inline src/controllers/userController.js:105]: Could you update what this method returns?

File: src/controllers/userController.js [hunk 11]
```diff
@@ -73,6 +73,62 @@ const logIn = async (req, res) => {
...
+ ******************************************************/
+
+const getAdmin = async (req, res) => {
```
Comment (in43sh) [inline src/controllers/userController.js:108]: I would change this controller name so that it reflects its purpose. Here's an alternative: `grantAdminRole`.

File: src/routes/mainRouter.js [hunk 12]
```diff
@@ -1,13 +1,26 @@
...
+const adminRole = require("../middlewares/adminRole");
 const router = express.Router();
 
 //user
 router.post("/signup", signupDataValidate, signUp);
-router.route("/login").post(loginDataValidate, logIn).get(authenticateUser);
+router
+  .route("/login")
+  .post(loginDataValidate, logIn)
+  .get(authenticateUser, getUser);
```
Comment (in43sh) [inline src/routes/mainRouter.js:20]: It's confusing to me that when I send a GET request to `/login`, I get the user info that is logged in. Maybe you could change it? An alternative is `/currentuser`.

**Review by in43sh — CHANGES_REQUESTED** [review]
Hi Bino! Thank you for implementing this important feature. It's working properly, but I would like you to address some changes that I put in this review. Let me know what you think about it.

File: src/middlewares/adminRole.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
+const userModel = require("../models/userSchema");
+const adminRole = async (req, res, next) => {
+  const { role, userId } = req.user;
+  try {
+    if (role !== "admin") {
```
Comment (ezbc) [inline src/middlewares/adminRole.js:5]: If I'm following the logic here correctly, any user can make a PUT request to update their own role to be an admin, is this correct? This would be insecure since any user could create their own account then assign themselves to be an admin.

In general I'd recommend that only admins can create admin users. It'd be good to add logic to this endpoint that the requesting user must have the admin role otherwise return a 403 forbidden.

But then you may ask, how do we create the first admin user? :eyes:  The first admin user can be created directly in the database. https://www.shellhacks.com/mongodb-create-user-database-admin-root/. You could either proceed with creating any other future admins in the database directly, or expose an endpoint for admin users to be able to create other admin users.

File: .env.example (hunk 3 above)
Comment (in43sh) [reply .env.example:3]: Also, I would replace `SECRET` with `JWT_SECRET` otherwise, it can be confused with hashing/salting secret.
